pages/en/zkapps/how-to-write-a-zkapp.mdx

```diff
@@ -474,6 +474,40 @@ super.deploy(args);
 If you don't need to add anything else to `deploy`, there's no need to override it, just leave it out.
 In the example above however, we are setting our state `x` to `Field(10)`. Uninitialized values will be zero.
 
+##### Custom data types
+
+As mentioned previously, smart contract `@method` arguments can be any of the built-in SnarkyJS types.
+
+However, what if you want to define your own data types?
+
+You can create your own custom data type for your smart contract using the `CircuitValue` class that SnarkyJS exposes. To start using the `CircuitValue` class to define your data types, you start by extending the `CircuitValue` class.
+Then similar to adding state to your smart contract, you can add properties to your custom data type using the `@prop` decorator.
+
+For example, let us say you want to create a custom data type called `Point` that represents a 2D point on a grid. You can do this by creating a new class that extends the `CircuitValue` class:
+
+```ts
+class Point extends CircuitValue {
+  @prop x: Field;
+  @prop y: Field;
+
+  // `fromPoint` is a method to create a `Point`
+  // from an `x` and `y` value that are typed as `number`.
+  static fromPoint(x: number, y: number) {
+    return new Point(Field(x), Field(y));
+  }
+}
+```
+
+Now that you have defined your custom data type, you can use it in your smart contract by adding a new state variable to your smart contract that uses your custom data type:
+
+```ts
+class Grid extends SmartContract {
+  @state(Point) p = State<Point>();
+}
+```
+
+`CircuitValues` can also be used as arguments for `@method`s in addition to SnarkyJS built-in types like `Field`, `Bool`, `UInt64`, etc. Note that `CircuitValues` themselves must be built of out SnarkyJS built-in types or other `CircuitValues`.
```

Comment:
> `CircuitValues` can also be used as arguments for `@method`s in addition to SnarkyJS built-in types like `Field`, `Bool`, `UInt64`, etc.

This feels redundant with your first sentence in this section. I wonder if it makes sense to just move your examples ", such as `Field`, `Bool`, `UInt64`, etc." to the end of your very first sentence in this section? 

Or to simply say:

> Note that your `CircuitValue`'s can contain SnarkyJS built-in types like `Field`, `Bool`, `UInt64`, etc _or_ even other custom types that you've defined which are based on the `CircuitValue` class.